### PR TITLE
[GH-100] feature/add-timestamped-speaker-turns

### LINKED ISSUE #100: Add timestamped speaker turns to transcript format
## Use Case
It would be nice if our transcript format includes `timestamped-speaker-turns` where each timestamped item in transcript.data is a speaker turn.

## Solution
seattlechannel.org sometime provides vtt caption files that we can use to create timestamped speaker turns.

To do this we need to:
1. Modify `SeattleEventScraper` to get the vtt file. Specifically, information about the vtt file can be found in the same place where information about the video_uri is found.
I'm looking at this particular line in `SeattleEventScraper`: 
`video = video_and_thumbnail.find("a").get("onclick")`
2. Add a caption_uri field to each scraped event. For example, caption_uri could be:
`https://seattlechannel.org/documents/seattlechannel/closedcaption/2019/brief_071519_2011955.vtt`
3. Add a new module that implements `SRModel` whose transcribe function will take the caption_uri and create `timestamped-speaker-turns` transcript.
transcribe function will use [webvtt-py](https://github.com/glut23/webvtt-py) to parse the vtt file and then leverage the fact that new speaker turn begins with `&gt;&gt;`
4. Modify `EventGatherPipeline` to use caption_uri and new SRModel to create `timestamped-speaker-turns` transcript, if caption_uri is valid. If caption_uri is invalid, use the other SRModel to create transcripts.

### PR DESCRIPTION
**Pull request by tohuynh**
Add `timestamped-speaker-turns` to transcript format.
Resolves #100 

- Modified `SeattleEventScraper` to scrape for closed caption file URI

- Added a new module `WebVTTSRModel` that implements `SRModel` that takes the closed caption file URI and produces transcripts in three formats: `raw`, `timestamped-sentences`, and `timestamped-speaker-turns`

- Modified `EventGatherPipeline` to use a mixture of speech recognition models. That is, if there is a closed caption file URI, use  `WebVTTSRModel`  to produce transcripts. Else, use `GoogleCloudSRModel` to produce transcripts.

Tests

- Added a few fake files(fake_caption.vtt, fake_timestamped_sentences.json) to test `WebVTTSRModel`

- Test `EventGatherPipeline` new mixture of speech recognition models. That is, test for when `WebVTTSRModel` failed and succeeded in producing transcripts.

- Added an example_transcript_speaker_turns.json to be consistent with other example_transcripts.json

### CODE REVIEW

File: cdptools/pipelines/event_gather_pipeline.py [hunk 1]
```diff
@@ -124,45 +146,62 @@ def task_transcript_get_or_create(self, key: str, audio_uri: str, phrases: Optio
...
+            # Setup temporary filenames, sorted by preference
+            tmp_filenames = [
+                f"{key}_ts_speaker_turns_transcript_0.json",
+                f"{key}_ts_sentences_transcript_0.json",
+                f"{key}_ts_words_transcript_0.json",
+                f"{key}_raw_transcript_0.json"
+            ]
+
+            # Iterate through tmp_filenames and check if any of them exists
+            main_transcript_uri = None
```
Comment (JacksonMaxfield) [inline cdptools/pipelines/event_gather_pipeline.py]: This is a better implementation. Nice job.

File: cdptools/sr_models/sr_model.py [hunk 2]
```diff
@@ -57,25 +58,28 @@ def wrap_and_format_transcript_data(
     @abstractmethod
     def transcribe(
         self,
-        audio_uri: Union[str, Path],
+        file_uri: Union[str, Path],
```
Comment (JacksonMaxfield) [inline cdptools/sr_models/sr_model.py]: Nice change to make move versatile.

File: cdptools/sr_models/webvtt_sr_model.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        with open(timestamped_sentences_save_path, "w") as write_out:
+            json.dump(timestamped_sentences, write_out)
+
+        with open(timestamped_speaker_turns_save_path, "w") as write_out:
+            json.dump(timestamped_speaker_turns, write_out)
+
+        # Return the save path
+        return SRModelOutputs(
+            raw_transcript_save_path,
+            1,
```
Comment (JacksonMaxfield) [inline cdptools/sr_models/webvtt_sr_model.py]: Can you make these both have the keyword arguments just for my own sanity and future-proofing safety.

File: cdptools/tests/data/example_event_pipeline_config_with_mixture_sr_model.json [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        "module_path": "cdptools.audio_splitters.ffmpeg_audio_splitter",
+        "object_name": "FFmpegAudioSplitter",
+        "object_kwargs": {}
+    },
+    "speech_recognition_model": {
+        "try": {
+            "module_path": "cdptools.sr_models.webvtt_sr_model",
+            "object_name": "WebVTTSRModel",
+            "object_kwargs": {
+                "new_turn_pattern": "&gt;"
```
Comment (JacksonMaxfield) [inline cdptools/tests/data/example_event_pipeline_config_with_mixture_sr_model.json]: Curious, shouldn't this be `&gt;&gt;`? I always saw two of them not a single.

File: cdptools/tests/pipelines/test_event_gather_pipeline.py [hunk 5]
```diff
@@ -106,6 +118,31 @@ def mocked_sr_model(
...
+
+    return mocked_model
+
+
+@pytest.fixture
+def mocked_webvtt_sr_model_with_request_exception() -> WebVTTSRModel:
+    mocked_model = mock.Mock(WebVTTSRModel("any-new-turn-pattern"))
+    # Mock RequestException for transcribe with invalid-caption-uri
+    mocked_model.transcribe.side_effect = RequestException("invalid-caption-uri")
+    return mocked_model
```
Comment (JacksonMaxfield) [inline cdptools/tests/pipelines/test_event_gather_pipeline.py]: Well done with adding this test!

File: docs/system_design.md [hunk 6]
```diff
@@ -89,6 +89,11 @@ transcripts for city council meetings, maybe you will write a module that simply
 already high quality transcript back, no need for any processing. These are both valid speech recognition models, it's
 simply that one of them produces transcripts that have 100% confidence.
 
+If your city provides closed caption files for city council meetings, you can write a module that takes the closed caption
+file and produce transcripts. If not all city council meetings have closed caption files, you can use a mixture of speech
+recognition models. That is, if there is a closed caption file, use that to produce transcripts. If there is no closed
+caption file you can use a `sr_models.GoogleCloudSRModel` to produce transcripts from an audio file.
```
Comment (JacksonMaxfield) [inline docs/system_design.md]: Great documentation already, maybe add a link or explain the `try`, `catch` of the config file as well.

File: docs/transcript_formats.md [hunk 7]
```diff
@@ -51,9 +52,39 @@ A middle ground between `format: raw` and `format: timestamped-words`. Like all,
...
+```json
+{
+    "format": "timestamped-speaker-turns",
+    "annotations": [
+        {}
+    ],
+    "confidence": 0.0,
+    "data": [
+        {
+            "speaker": "speaker-0",
```
Comment (JacksonMaxfield) [inline docs/transcript_formats.md]: I would add a comment that you can leave this as an empty string if it is unknown.

File: cdptools/pipelines/event_gather_pipeline.py (hunk 1 above)
Comment (tohuynh) [reply cdptools/pipelines/event_gather_pipeline.py]: Yea, a nested block of 4 try/exceptions is a bit too hard to read/understand.

File: cdptools/tests/data/example_event_pipeline_config_with_mixture_sr_model.json (hunk 4 above)
Comment (tohuynh) [reply cdptools/tests/data/example_event_pipeline_config_with_mixture_sr_model.json]: Sometimes in the caption file it's `&gt;&gt;` or `&gt;&gt;&gt;`. I couldn't find a way to write a '2 or more' regexp in line 31 of WebVTTSRModel.py `self.new_turn_pattern = r"^({})+\s*(.+)$".format(new_turn_pattern)`. Do you know a way?

I could change it to `"new_turn_pattern": "&gt;&gt;"`, but how would I change the code for line 31?

File: cdptools/sr_models/webvtt_sr_model.py (hunk 3 above)
Comment (tohuynh) [reply cdptools/sr_models/webvtt_sr_model.py]: Yes

File: cdptools/tests/data/example_event_pipeline_config_with_mixture_sr_model.json (hunk 4 above)
Comment (JacksonMaxfield) [reply cdptools/tests/data/example_event_pipeline_config_with_mixture_sr_model.json]: Ohhhh I see how you are using it. Makes sense that it's just a boolean search.

**Comment by tohuynh** [comment]
New changes has those changes you requested.

**Pull request merged**